pkg/backends/calico/routes.go

```diff
@@ -15,30 +15,240 @@ package calico
+	epHandler := cache.ResourceEventHandlerFuncs{AddFunc: rg.onEPAdd, UpdateFunc: rg.onEPUpdate, DeleteFunc: rg.onEPDelete}
+	rg.epIndexer, rg.epInformer = cache.NewIndexerInformer(epWatcher, &v1.Endpoints{}, 0, epHandler, cache.Indexers{})
+
+	log.Debug("-NewRouteGenerator")
 	return
 }
 
 func (rg *routeGenerator) Start() (err error) {
 	// MVP implementation: read CIDRs to advertise,
 	// comma-separated, from an environment variable
 	// CALICO_STATIC_ROUTES.
-	routeString := os.Getenv("CALICO_STATIC_ROUTES")
-	cidrs := []string{}
-	for _, route := range strings.Split(routeString, ",") {
-		cidr := strings.TrimSpace(route)
-		if cidr != "" {
-			cidrs = append(cidrs, cidr)
+	ch := make(chan struct{})
+	go rg.svcInformer.Run(ch)
+	go rg.epInformer.Run(ch)
+
+	/*
+		routeString := os.Getenv("CALICO_STATIC_ROUTES")
+		cidrs := []string{}
+		for _, route := range strings.Split(routeString, ",") {
+			cidr := strings.TrimSpace(route)
+			if cidr != "" {
+				cidrs = append(cidrs, cidr)
+			}
+		}
+		rg.client.updateRoutes(cidrs)
+	*/
+	return
+}
+
+// setRouteForSvc handles the main logic to check if a specified service or endpoint
+// should have its route advertised by the node running this code
+func (rg *routeGenerator) setRouteForSvc(svc *v1.Service, ep *v1.Endpoints) {
+	log.Debug("+setRouteForSvc")
+	rg.Lock()
+	defer rg.Unlock()
+
+	// ensure both are not nil
+	if svc == nil && ep == nil {
+		log.Warn("setRouteForSvc: both service and endpoint cannot be nil, passing...")
+		return
+	}
+
+	var (
+		key string
+		err error
+	)
+
+	if svc == nil {
+		// ep received but svc nil
+		// get key
+		key, err = cache.MetaNamespaceKeyFunc(ep)
+		if err != nil {
+			log.WithField("svc", svc.Name).WithError(err).Warn("setRouteForSvc: error on retrieving key for endpoint, passing")
+			return
+		}
+		// get svc
+		svcIface, exists, err := rg.svcIndexer.GetByKey(key)
+		if err != nil {
+			log.WithField("key", key).WithError(err).Warn("setRouteForSvc: error on retrieving service for key, passing")
+			return
+		} else if !exists {
+			log.WithField("key", key).Warn("setRouteForSvc: service for key not found, passing")
+			return
+		}
+		svc = svcIface.(*v1.Service)
+	} else if ep == nil {
+		// svc received but ep nil
+		// get key
+		key, err = cache.MetaNamespaceKeyFunc(svc)
+		if err != nil {
+			log.WithField("ep", ep.Name).WithError(err).Warn("setRouteForSvc: error on retrieving key for service, passing")
+			return
+		}
+		// get ep
+		epIface, exists, err := rg.epIndexer.GetByKey(key)
+		if err != nil {
+			log.WithField("key", key).WithError(err).Warn("setRouteForSvc: error on retrieving endpoint for key, passing")
+			return
+		} else if !exists {
+			log.WithField("key", key).Warn("setRouteForSvc: service for endpoint not found, passing")
+			return
+		}
+		ep = epIface.(*v1.Endpoints)
+	}
+
+	// always set if externalTrafficPolicy == cluster
```

Comment:
This function is a bit large, I think it would be really nice to compartmentalize bits of it which will make it easier to read and also to write UTs for.

For example, I think we can pull out this logic here into a separate function, something like this:

```go
// Returns true if this service should be advertised on this node, false otherwise. 
func advertiseThisService(svc, *v1.Service, ep *v1.Endpoints) bool {}
```